### PR TITLE
Add message and chat models

### PR DESCRIPTION
**Pull request by antoalconti**
- Message model and spec was added
- Chat model and spec was added

### CODE REVIEW

File: app/models/chat.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class Chat < ApplicationRecord
+  belongs_to :user_a,  class_name: 'User', foreign_key: 'user_a_id'
```
Comment (TimoPeraza) [inline app/models/chat.rb]: maybe, :user_receiver && :user_sender could be better names, WDYT?

File: app/models/chat.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+  belongs_to :user_b,  class_name: 'User', foreign_key: 'user_b_id'
+  has_many :messages
```
Comment (TimoPeraza) [inline app/models/chat.rb]: maybe here, you could add a ```dependent: :destroy```

File: app/models/message.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+class Message < ApplicationRecord
+  belongs_to :user
+  belongs_to :chat
+
+  validates :text, presence: true
+  validate :user_blongs_to_chat, on: :create
+
+  def user_blongs_to_chat
```
Comment (TimoPeraza) [inline app/models/message.rb]: maybe this method should be private

File: app/models/user.rb [hunk 4]
```diff
@@ -11,4 +11,8 @@ class User < ApplicationRecord
 
   validates :full_name, presence: true
   validates :gender, presence: true, inclusion: { in: genders.keys }
+
+  def chats
+    Chat.where(user_a_id: id) + Chat.where(user_b_id: id)
```
Comment (santir489) [inline app/models/user.rb]: Maybe you can combine this two queries, so you make only one request to DB.. WDYT?

File: spec/models/message_spec.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
+require 'rails_helper'
+
+RSpec.describe Message, type: :model do
+  subject { create(:message) }
+
+  describe 'validations' do
+    it { should validate_presence_of(:text) }
+
+    describe 'user does not blongs to chat users' do
```
Comment (santir489) [inline spec/models/message_spec.rb]: `belong`

File: spec/models/message_spec.rb [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+      let(:user_a) { create(:user) }
+      let(:user_b) { create(:user) }
+      let!(:chat_a) { create(:chat, user_a: user_a) }
+      let(:test) { create(:message, chat: chat_a, user: user_b) }
+
+      it 'does not create message' do
+        msg = 'Validation failed: Message User must be the same than one of the chat users'
+        expect { test }.to raise_error(ActiveRecord::RecordInvalid, msg)
+      end
```
Comment (santir489) [inline spec/models/message_spec.rb]: mmm are you not using subject here ?

maybe you can do something like 
`subject { build(:message, chat: chat_a, user: user_b) } `
and then validate the following
```
subject.valid?
expect(subject.errors.messages).to include(error_message)
```

WDYT?

File: app/models/chat.rb (hunk 1 above)
Comment (antoalconti) [reply app/models/chat.rb]: In the chat model, there is not a sender or receiver both of them send and receive messages.

File: app/models/user.rb (hunk 4 above)
Comment (antoalconti) [reply app/models/user.rb]: Yes, I think is the optimal way to do it!

File: app/models/chat.rb (hunk 1 above)
Comment (sebaherrera07) [reply app/models/chat.rb]: Agree with anto, there is no sender or receiver in the chat.

File: app/models/message.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+  validates :text, presence: true
+  validate :user_belongs_to_chat, on: :create
+
+  private
+
+  def user_belongs_to_chat
+    return if chat.user_a == user || chat.user_b == user
+
+    errors.add(:message, I18n.t('api.models.message.errors.chat_diferent_users'))
```
Comment (sebaherrera07) [inline app/models/message.rb]: You missed a `f` : `chat_different_users `

File: config/locales/en.yml [hunk 8]
```diff
@@ -10,3 +10,6 @@ en:
       target:
         errors:
           reached_limit: 'You reached the %{max} limit of targets! Delete one if you want to create another.'
+      message:
+        errors:
+          chat_diferent_users: 'User must be the same than one of the chat users'
```
Comment (sebaherrera07) [inline config/locales/en.yml]: I would say: `User must be one of the chat users`.

File: spec/models/chat_spec.rb [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+RSpec.describe Chat, type: :model do
+  subject { create(:chat) }
+
+  describe 'associations' do
+    it { should have_many(:messages) }
+    it { should belong_to(:user_a) }
+    it { should belong_to(:user_b) }
+  end
+
+  describe '.create' do
```
Comment (sebaherrera07) [inline spec/models/chat_spec.rb]: This test shouldné be necessary.

File: spec/models/message_spec.rb [hunk 10]
```diff
@@ -0,0 +1,28 @@
+require 'rails_helper'
+
+RSpec.describe Message, type: :model do
+  subject { create(:message) }
+
+  describe 'validations' do
+    it { should validate_presence_of(:text) }
+
+    describe 'user does not belongs to chat users' do
```
Comment (sebaherrera07) [inline spec/models/message_spec.rb]: I would put the name of the validation method: `'#user_belongs_to_chat'`

File: spec/models/message_spec.rb [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+      let(:user_a) { create(:user) }
+      let(:user_b) { create(:user) }
+      let!(:chat_a) { create(:chat, user_a: user_a) }
+      let(:test) { create(:message, chat: chat_a, user: user_b) }
+      subject { build(:message, chat: chat_a, user: user_b) }
+      it 'does not create message' do
```
Comment (sebaherrera07) [inline spec/models/message_spec.rb]: You should validate 3 cases:
- Is user A
- Is user B
- Is none of the users.

File: spec/models/message_spec.rb [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+        error_message = I18n.t('api.models.message.errors.chat_diferent_users')
+        subject { build(:message, chat: chat_a, user: user_b) }
```
Comment (sebaherrera07) [inline spec/models/message_spec.rb]: You defined this subject twice.

File: spec/models/user_spec.rb [hunk 13]
```diff
@@ -12,4 +12,13 @@
   describe 'associations' do
     it { should have_many(:targets) }
   end
+
+  describe '#chats' do
+    it 'should return chats collections' do
+      user_a = create(:user)
```
Comment (sebaherrera07) [inline spec/models/user_spec.rb]: I would suggest creating this user outside the `it` block, and inside the `describe` block.

File: spec/models/user_spec.rb [hunk 14]
```diff
@@ -12,4 +12,13 @@
...
+      chat_a = create(:chat, user_a: user_a)
```
Comment (sebaherrera07) [inline spec/models/user_spec.rb]: I would suggest creating this chats outside the it block, and inside the describe block.

File: spec/models/message_spec.rb [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+
+  describe 'validations' do
+    it { should validate_presence_of(:text) }
+
+    describe '#user_belongs_to_chat' do
+      let!(:user_a) { create(:user) }
+      let!(:user_b) { create(:user) }
+      let!(:chat_a) { create(:chat, user_a: user_a, user_b: user_b) }
+
+      describe 'when user is no one of chat users' do
```
Comment (sebaherrera07) [inline spec/models/message_spec.rb]: This should be a context

File: spec/models/message_spec.rb [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+        subject { build(:message, chat: chat_a, user: user) }
+
+        it 'does not create message whit user_b' do
+          error_message = I18n.t('api.models.message.errors.chat_different_users')
+          subject.valid?
+          expect(subject.errors.messages).to include(message: [error_message])
+        end
+      end
+
+      describe 'when user is user_a of chat users' do
```
Comment (sebaherrera07) [inline spec/models/message_spec.rb]: This should be a context

File: spec/models/message_spec.rb [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+        subject { create(:message, chat: chat_a, user: user_a) }
+
+        it 'creates a message' do
+          expect { subject }.to change { Message.count }.by(1)
+        end
+      end
+
+      describe 'when user is user_b of chat users' do
```
Comment (sebaherrera07) [inline spec/models/message_spec.rb]: This should be a context

File: spec/models/message_spec.rb [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+    describe '#user_belongs_to_chat' do
+      let!(:user_a) { create(:user) }
+      let!(:user_b) { create(:user) }
+      let!(:chat_a) { create(:chat, user_a: user_a, user_b: user_b) }
+
+      describe 'when user is no one of chat users' do
+        let(:user) { create(:user) }
+        subject { build(:message, chat: chat_a, user: user) }
+
+        it 'does not create message whit user_b' do
```
Comment (sebaherrera07) [inline spec/models/message_spec.rb]: with

File: spec/models/user_spec.rb [hunk 19]
```diff
@@ -12,4 +12,13 @@
   describe 'associations' do
     it { should have_many(:targets) }
   end
+
+  describe '#chats' do
+    let(:user_a) { create(:user) }
```
Comment (sebaherrera07) [inline spec/models/user_spec.rb]: I think you can just call it `user` instead of `user_a` . Because then you use it as both `user_a` and `user_b`

**Review by sebaherrera07 — APPROVED** [review]
Excellent job! 🚀

File: spec/models/user_spec.rb [hunk 20]
```diff
@@ -12,4 +12,13 @@
   describe 'associations' do
     it { should have_many(:targets) }
   end
+
+  describe '#chats' do
+    let(:user) { create(:user) }
+    let(:chat_a) { create(:chat, user_a: user) }
+    let(:chat_b) { create(:chat, user_b: user) }
+    it 'should return chats collections' do
```
Comment (santir489) [inline spec/models/user_spec.rb]: Leave empty line between `let` and `it` blocks

**Review by santir489 — APPROVED** [review]
👏 👏

**Pull request merged**